### PR TITLE
Detect new files under known paths in filestream input

### PR DESCRIPTION
**Pull request by kvch**
## What does this PR do?

This PR fixes the `FileWatcher` of `filestream` input. Now a file is considered new even if the scanner has already found it in the previous iteration and the underlying file is different.

In the PR the file comparator function is passed as a parameter to make unit testing easier.

## Why is it important?

The problem is if an input file is renamed and a new file shows up, Filebeat did not register is as a new file. The new file was either considered updated. Or if the new file was smaller than the previous file, the file was deemed truncated and the complete contents of the previous file was reread from the beginning.

The issue was reported initially on Discuss: https://discuss.elastic.co/t/filebeat-filestream-input-rereading-rotated-log-files/300038

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: filebeat/input/filestream/fswatch_test.go [hunk 1]
```diff
@@ -118,6 +119,58 @@ func setupFilesForScannerTest(t *testing.T, tmpDir string) {
...
+		sameFileFunc: os.SameFile,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go w.Run(ctx)
+
+	appLogPath := filepath.Join(tmpDir, "app.log")
+	rotatedAppLogPath := filepath.Join(tmpDir, "app.log.1")
+	err = ioutil.WriteFile(appLogPath, []byte("my longer log line"), 0o600)
```
Comment (belimawr) [inline filebeat/input/filestream/fswatch_test.go:143]: `ioutil` has been deprecated, you can use `os.WriteFile` instead:
```suggestion
	err = os.WriteFile(appLogPath, []byte("my longer log line"), 0o600)
```

File: filebeat/input/filestream/fswatch_test.go [hunk 2]
```diff
@@ -118,6 +119,58 @@ func setupFilesForScannerTest(t *testing.T, tmpDir string) {
...
+	require.Equal(t, evt.Op, loginp.OpCreate, "new file should be detected")
+	require.Equal(t, evt.OldPath, "", "new file does not have an old path set")
+	require.Equal(t, evt.NewPath, appLogPath, "new file does not have an old path set")
+
+	err = os.Rename(appLogPath, rotatedAppLogPath)
+	if err != nil {
+		t.Fatalf("failed to rotate active file: %v", err)
+	}
+
+	err = ioutil.WriteFile(appLogPath, []byte("shorter line"), 0o600)
```
Comment (belimawr) [inline filebeat/input/filestream/fswatch_test.go:158]: Same thing here
```suggestion
	err = os.WriteFile(appLogPath, []byte("shorter line"), 0o600)
```

File: filebeat/input/filestream/fswatch_test.go [hunk 3]
```diff
@@ -118,6 +119,58 @@ func setupFilesForScannerTest(t *testing.T, tmpDir string) {
...
+	rotatedAppLogPath := filepath.Join(tmpDir, "app.log.1")
+	err = ioutil.WriteFile(appLogPath, []byte("my longer log line"), 0o600)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	evt := w.Event()
+	require.Equal(t, evt.Op, loginp.OpCreate, "new file should be detected")
+	require.Equal(t, evt.OldPath, "", "new file does not have an old path set")
+	require.Equal(t, evt.NewPath, appLogPath, "new file does not have an old path set")
```
Comment (belimawr) [inline filebeat/input/filestream/fswatch_test.go:151]: It's a detail, but the signature of [`requre.Equal`](https://pkg.go.dev/github.com/stretchr/testify/require#Equal) is:

```
func Equal(t TestingT, expected interface{}, actual interface{}, msgAndArgs ...interface{})
```

The expected value comes first. This inversion will make the test error harder to understand.
```suggestion
	require.Equal(t, loginp.OpCreate, evt.Op, "new file should be detected")
	require.Equal(t, "", evt.OldPath, "new file does not have an old path set")
	require.Equal(t, appLogPath, evt.NewPath, "new file does not have an old path set")
```

The same applies to the other similar blocks on this function.

**Review by belimawr — CHANGES_REQUESTED** [review]
Looks good, just some small details in the tests.

**Review by belimawr — APPROVED** [review]
:tada:

**Comment by kvch** [comment]
The test is not passing on Windows for some reason. So I moved the test to non-windows tests for now. I am comfortable with merging this because we are not breaking anything for Windows users. We just still do not support something that was broken in the first place. The test failure needs more investigation on Windows.

**Pull request merged**